# UDACITY PROJECT 5 - FINAL PROJECT
## Data Exploration - Ford GoBike System Data Analysis
### *Jhonatan Nagasako*
#### *28-FEB-2021*

<hr size="5"/>

<a id='contents'></a>
# Table of Contents (click link to section)

<a href="#intro">A. INTRODUCTION</a><br><br>
<a href="#gather">1. GATHERING DATA</a><br><br>
<a href="#assess">2. ASSESSING DATA</a><br>
* 2.1 <a href="#todo">[Cleaning To-Do List]</a><br>

<a href="#clean">3. CLEANING DATA</a><br>

<a href="#store">4. STORING DATA</a><br>

<a href="#explore">5. DATA EXPLORATION</a>    
* 5.1 <a href="#preliminary1">[Exporation - Preliminary Review]</a><br>
* 5.2 <a href="#univariate1">[Exporation - Univariate Exploration]</a><br>
* 5.3 <a href="#bivariate1">[Exporation - Bivariate Exploration]</a><br>
* 5.4 <a href="#multivariate1">[Exporation - Multivariate Exploration]</a><br>

<a href="#discussion">6. DISCUSSION</a><br> 
* 6.1 <a href="#preliminary2">[Discussion - Preliminary Review]</a><br>
* 6.2 <a href="#univariate2">[Discussion - Univariate Exploration]</a><br>
* 6.3 <a href="#bivariate2">[Discussion - Bivariate Exploration]</a><br>
* 6.4 <a href="#multivariate2">[Discussion - Multivariate Exploration]</a><br>  

<a href="#conclusion">7. CONCLUSION</a>



<hr size="5"/>

<a id='intro'></a>
# A. INTRODUCTION

This notebook will focus on the **data EXPLORATION** of Ford GoBike System. Data source can be found via [Udacity provided link to Google Docs/Drive](https://docs.google.com/document/d/e/2PACX-1vQmkX4iOT6Rcrin42vslquX2_wQCjIa_hbwD0xmxrERPSOJYDtpNc_3wwK_p9_KpOsfA6QVyEHdxxq7/pub?embedded=True). 

This data set includes information about individual rides made in a bike-sharing system covering the greater San Francisco Bay area.
* Note that this dataset will require some data wrangling in order to make it tidy for analysis. There are multiple cities covered by the linked system, and multiple data files will need to be joined together if a full year’s coverage is desired.
* Depending on scope of questions explored, additional data sources from other cities may be explored. Data can be accessed via [this page](https://www.google.com/url?q=https://www.bikeshare.com/data/&sa=D&source=editors&ust=1614518054096000&usg=AOvVaw38y_cueV0lTerb59CY7YsD) or [this page](https://www.google.com/url?q=https://github.com/BetaNYC/Bike-Share-Data-Best-Practices/wiki/Bike-Share-Data-Systems&sa=D&source=editors&ust=1614518054097000&usg=AOvVaw2OskG9ApXPoPZlezrpwmXp).

<a href="#contents">[Table of Contents]</a>

<hr size="5"/>
<h5><center>📚 Gathering START -- Project START 📚</center></h5>        
<hr size="5"/>

<a id='gather'></a>
# 1. GATHERING DATA

<font color=blue>

<a href="#contents">[Table of Contents]</a>

In [1]:
# import statements for all of the packages used for analysis

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
import statsmodels.api as sm;

# Remember to include a 'magic word' so that your visualizations are plotted
#   inline with the notebook. See this page for more:
#   http://ipython.readthedocs.io/en/stable/interactive/magics.html

%matplotlib inline

In [2]:
# gather .csv file
df = pd.read_csv('201902-fordgobike-tripdata.csv')
df.head(3)

duration_sec                start_time                  end_time  \
0         52185  2019-02-28 17:32:10.1450  2019-03-01 08:01:55.9750   
1         42521  2019-02-28 18:53:21.7890  2019-03-01 06:42:03.0560   
2         61854  2019-02-28 12:13:13.2180  2019-03-01 05:24:08.1460   

   start_station_id                                start_station_name  \
0              21.0  Montgomery St BART Station (Market St at 2nd St)   
1              23.0                     The Embarcadero at Steuart St   
2              86.0                           Market St at Dolores St   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.789625              -122.400811            13.0   
1               37.791464              -122.391034            81.0   
2               37.769305              -122.426826             3.0   

                               end_station_name  end_station_latitude  \
0                Commercial St at Montgomery St             37.794231   
1                            Berry St at 4th St             37.775880   
2  Powell St BART Station (Market St at 4th St)             37.786375   

   end_station_longitude  bike_id user_type  member_birth_year member_gender  \
0            -122.402923     4902  Customer             1984.0          Male   
1            -122.393170     2535  Customer                NaN           NaN   
2            -122.404904     5905  Customer             1972.0          Male   

  bike_share_for_all_trip  
0                      No  
1                      No  
2                      No

In [3]:
# high-level overview of data shape and composition
print('\n----Shape of File (row, column)---\n')
print(df.shape)
print('\n----Data Types---\n')
print(df.dtypes)
print('\n----Unique Values---\n')
print(df.nunique())
print('\n----Number of missing values---\n')
print(df.isnull().sum())


----Shape of File (row, column)---

(183412, 16)

----Data Types---

duration_sec                 int64
start_time                  object
end_time                    object
start_station_id           float64
start_station_name          object
start_station_latitude     float64
start_station_longitude    float64
end_station_id             float64
end_station_name            object
end_station_latitude       float64
end_station_longitude      float64
bike_id                      int64
user_type                   object
member_birth_year          float64
member_gender               object
bike_share_for_all_trip     object
dtype: object

----Unique Values---

duration_sec                 4752
start_time                 183401
end_time                   183397
start_station_id              329
start_station_name            329
start_station_latitude        334
start_station_longitude       335
end_station_id                329
end_station_name              329
end_station_latitude       

<hr size="5"/>
<h5><center>🔎 Gathering END ➜ Assessing START 🔎</center></h5>        
<hr size="5"/>

<a id='assess'></a>
# 2. ASSESSING DATA (Cleaning To-Do List)

<font color=blue>

<a href="#contents">[Table of Contents]</a>

<font color='red'>

<a id='todo'></a>
## QUALITY (click hyperlink question to go to section in notebook!)

<a href="#quality-section">[Go to QUALITY section]</a>
    
✔️ 1. <a href="#Q1">[Remove rows with ```NULL``` in ```end_station_id``` column (197 items)]</a>
    
    
2. <a href="#Q2">[Item above may fix this, but remove rows with ```NULL``` in thee following columns (197 items each)]</a>

    * start_station_id
    * start_station_name
    * end_station_name
    
    
3. <a href="#Q3">[Remove rows in ```start_station_latitude``` in ```start_station_longitude``` columns]</a>
    
    * should change from 334-latitude and 335-longitude to 329 items
    
    
4. <a href="#Q4">[Review ```member_birth_year``` and determine good cut-off date--remove rows respectively]</a>
    
    * E.g., There is someone born in 1900 that rented a bike in 2019... that makes that person over 100 years old!
    * consider average age and oldest person... this could be errors that can be removed

    
5. <a href="#Q5">[Convert ```duration_sec``` to minutes, hours, or/and days]</a>
    
<br>
    
>**Tips for Common Data Quality Issues**
>1. Missing data
2. Invalide data (e.g., state a negative height, or other datatype validation errors--str vs int vs float, think there can only be 2 people in a room... not 2.54 people in a room... *unless there's ghosts lol*)
3. Inaccurate data (e.g., specifying a foot = 5 inches, which is WRONG. A foot = 12 inches)
4. Inconsistent data (e.g., mixing up units, some data captured as cm instead of inches)



## TIDINESS
    
<a href="#tidy-section">[Go to TIDY section]</a>
    
1. <a href="#T1">[Create columns to break apart ```start_time``` to ```date``` and ```time```]</a>

    
2. <a href="#T2">[Create columns to break apart ```end_time``` to ```date``` and ```time```]</a>

<br>
    
>**Tips for Tidying**
>1. Each variable forms a column.
2. Each observation forms a row.
3. Each type of observational unit forms a table
*Reference for [tidy data here](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html)*

<br>
    
## FEATURE ENGINEEERING
1. <a href="#F1">[Create column for ```day of week``` for the start/end days--to see what days bikes are rented the most]</a>
    
<a href="#assess">[Assessing Data Requirements]</a> <a href="#contents">[Table of Contents]</a>

<hr size="5"/>
<h5><center>🧹 Assessing END ➜ Cleaning START 🧹</center></h5>        
<hr size="5"/>

<a id='clean'></a>
# 3. CLEANING DATA

<font color=blue>
    
   
<a href="#todo">[Cleaning and Tidying To-do List]</a> 
<a href="#contents">[Table of Contents]</a>

<a id='quality-section'></a>

<font color='red'>
    
## QUALITY ISSUES ADDRESSED -- note that ```df``` will change to ```dfc``` to indicate cleaned data
    
<a href="#todo">[Cleaning and Tidying To-do List (from Section 2: Assessing data]</a>

<a id='Q1'></a>

<font color='red'>

### ✔️ 1. <a href="#Q1">[Remove rows with ```NULL``` in ```end_station_id``` column (197 items)]</a>

<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [4]:
# before cleaning 
print('\n----BEFORE CLEANING >> Number of missing values---\n')
print(df.isnull().sum())
print('\n----BEFORE CLEANING >> Shape of File (row, column)---\n')
print(df.shape)


----BEFORE CLEANING >> Number of missing values---

duration_sec                  0
start_time                    0
end_time                      0
start_station_id            197
start_station_name          197
start_station_latitude        0
start_station_longitude       0
end_station_id              197
end_station_name            197
end_station_latitude          0
end_station_longitude         0
bike_id                       0
user_type                     0
member_birth_year          8265
member_gender              8265
bike_share_for_all_trip       0
dtype: int64

----BEFORE CLEANING >> Shape of File (row, column)---

(183412, 16)


In [5]:
# create test df for prototyping
test = df

# remove missing values
test.dropna(subset=['end_station_id'], inplace=True)
print('\n----AFTER CLEANING >> Number of missing values---\n')
print(df.isnull().sum())
print('\n----AFTER CLEANING >> Shape of File (row, column)---\n')
print(df.shape)


----AFTER CLEANING >> Number of missing values---

duration_sec                  0
start_time                    0
end_time                      0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
bike_id                       0
user_type                     0
member_birth_year          8263
member_gender              8263
bike_share_for_all_trip       0
dtype: int64

----AFTER CLEANING >> Shape of File (row, column)---

(183215, 16)


In [6]:
# reassign cleaned data to dfc
dfc = test
print(dfc.shape)
dfc.head(3)

(183215, 16)


duration_sec                start_time                  end_time  \
0         52185  2019-02-28 17:32:10.1450  2019-03-01 08:01:55.9750   
1         42521  2019-02-28 18:53:21.7890  2019-03-01 06:42:03.0560   
2         61854  2019-02-28 12:13:13.2180  2019-03-01 05:24:08.1460   

   start_station_id                                start_station_name  \
0              21.0  Montgomery St BART Station (Market St at 2nd St)   
1              23.0                     The Embarcadero at Steuart St   
2              86.0                           Market St at Dolores St   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.789625              -122.400811            13.0   
1               37.791464              -122.391034            81.0   
2               37.769305              -122.426826             3.0   

                               end_station_name  end_station_latitude  \
0                Commercial St at Montgomery St             37.794231   
1                            Berry St at 4th St             37.775880   
2  Powell St BART Station (Market St at 4th St)             37.786375   

   end_station_longitude  bike_id user_type  member_birth_year member_gender  \
0            -122.402923     4902  Customer             1984.0          Male   
1            -122.393170     2535  Customer                NaN           NaN   
2            -122.404904     5905  Customer             1972.0          Male   

  bike_share_for_all_trip  
0                      No  
1                      No  
2                      No

*This process also removed the rows of missing information that was in the following columns:*

   * ```start_station_id```
   * ```start_station_name```
   * ```end_station_name```

<a id='Q2'></a>

<font color='red'>

### ✔️ 2. <a href="#Q2">[Item above may fix this, but remove rows with ```NULL``` in thee following columns (197 items each)]</a>

    * start_station_id
    * start_station_name
    * end_station_name

<a href="#todo">[Cleaning and Tidying To-do List]</a>

*See section <a href="#Q1">(Quality Question Task Q1)</a> for the removal of the missing values.*

<a id='Q3'></a>

<font color='red'>

### ✔️ 3. <a href="#Q3">[Remove rows in ```start_station_latitude``` in ```start_station_longitude``` columns]</a>
    
    * should change from 334-latitude and 335-longitude to 329 items

<a href="#todo">[Cleaning and Tidying To-do List]</a>

*See section <a href="#Q1">(Quality Question Task Q1)</a> for the removal of the missing values.*

<a id='Q4'></a>

<font color='red'>

### ✔️ 4. <a href="#Q4">[Review ```member_birth_year``` and determine good cut-off date--remove rows respectively]</a>

<a href="#todo">[Cleaning and Tidying To-do List]</a>

In [7]:
# high-level overview of data shape and composition
print('\n----INITIAL STATE > Shape of File (row, column)---\n')
print(dfc.shape)
print('\n----INITIAL STATE > Data Types---\n')
print(dfc.dtypes)
print('\n----INITIAL STATE > Unique Values---\n')
print(dfc.nunique())
print('\n----INITIAL STATE > Number of missing values---\n')
print(dfc.isnull().sum())


----INITIAL STATE > Shape of File (row, column)---

(183215, 16)

----INITIAL STATE > Data Types---

duration_sec                 int64
start_time                  object
end_time                    object
start_station_id           float64
start_station_name          object
start_station_latitude     float64
start_station_longitude    float64
end_station_id             float64
end_station_name            object
end_station_latitude       float64
end_station_longitude      float64
bike_id                      int64
user_type                   object
member_birth_year          float64
member_gender               object
bike_share_for_all_trip     object
dtype: object

----INITIAL STATE > Unique Values---

duration_sec                 4749
start_time                 183204
end_time                   183200
start_station_id              329
start_station_name            329
start_station_latitude        329
start_station_longitude       329
end_station_id                329
end_station_n

<a id='zero'></a>

In [8]:
# create test df for prototyping
test = dfc

# need to fill blank locations in order to change the data type
test = test.fillna(0)

# change member_birth_year from FLOAT to INT
test['member_birth_year'] = test['member_birth_year'].astype(int) 

print('\n----CLEANED STATE > Shape of File (row, column)---\n')
print(test.shape)
print('\n----CLEANED STATE > Data Types---\n')
print(test.dtypes)
print('\n----CLEANED STATE > Unique Values---\n')
print(test.nunique())
print('\n----CLEANED STATE > Number of missing values---\n')
print(test.isnull().sum())


----CLEANED STATE > Shape of File (row, column)---

(183215, 16)

----CLEANED STATE > Data Types---

duration_sec                 int64
start_time                  object
end_time                    object
start_station_id           float64
start_station_name          object
start_station_latitude     float64
start_station_longitude    float64
end_station_id             float64
end_station_name            object
end_station_latitude       float64
end_station_longitude      float64
bike_id                      int64
user_type                   object
member_birth_year            int32
member_gender               object
bike_share_for_all_trip     object
dtype: object

----CLEANED STATE > Unique Values---

duration_sec                 4749
start_time                 183204
end_time                   183200
start_station_id              329
start_station_name            329
start_station_latitude        329
start_station_longitude       329
end_station_id                329
end_station_n

*Removing the missing values in the ```member_birth_year``` also removed missing values in ```member_gender```*

Determine good cutoff range, note that the data set is from 02-FEB-2019.

* [Oldest person alive](https://www.google.com/search?q=oldest+person+alive&rlz=1C1GCEB_enUS893US893&oq=oldest+person&aqs=chrome.0.0i433j69i57j0l2j0i67j0l5.3255j0j7&sourceid=chrome&ie=UTF-8) is **118**

After the review of the [2010 U.S. Census Bureu](https://www.census.gov/prod/cen2010/briefs/c2010br-03.pdf), lets make *age 100* the cut off point when considering age and gender.

In [9]:
# create column for age (based on 2019)
test['age'] = 2019 - test['member_birth_year']

# help: https://datascience.stackexchange.com/questions/38720/shifting-the-last-column-in-the-dataframe-to-the-first-place
# move last column to first temporarily
cols = list(test.columns)
cols = [cols[-1]] + cols[:-1]
test = test[cols]

#display data before filtering
test1 = test
print('BEFORE FILTER - Basic Stats on AGE\n', test.age.describe())
print('\nBEFORE FILTER, Zero Count = ',test1.age.isin([0]).sum())
test.head(2)

BEFORE FILTER - Basic Stats on AGE
 count    183215.000000
mean        123.711525
std         412.013105
min          18.000000
25%          27.000000
50%          32.000000
75%          40.000000
max        2019.000000
Name: age, dtype: float64

BEFORE FILTER, Zero Count =  0


age  duration_sec                start_time                  end_time  \
0    35         52185  2019-02-28 17:32:10.1450  2019-03-01 08:01:55.9750   
1  2019         42521  2019-02-28 18:53:21.7890  2019-03-01 06:42:03.0560   

   start_station_id                                start_station_name  \
0              21.0  Montgomery St BART Station (Market St at 2nd St)   
1              23.0                     The Embarcadero at Steuart St   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.789625              -122.400811            13.0   
1               37.791464              -122.391034            81.0   

                 end_station_name  end_station_latitude  \
0  Commercial St at Montgomery St             37.794231   
1              Berry St at 4th St             37.775880   

   end_station_longitude  bike_id user_type  member_birth_year member_gender  \
0            -122.402923     4902  Customer               1984          Male   
1            -122.393170     2535  Customer                  0             0   

  bike_share_for_all_trip  
0                      No  
1                      No

In [10]:
# replace any values greater than 100 with zeros -- data set should remain the same
test['age'].values[test['age'].values > 100] = 0

# review age statitics distrubtion
print('AFTER FILTER - Basic Stats on AGE\n', test.age.describe())
print('\nAFTER FILTER, Zero Count in AGE column = ',test1.age.isin([0]).sum())
test.head(2)

AFTER FILTER - Basic Stats on AGE
 count    183215.000000
mean         32.607909
std          12.067822
min           0.000000
25%          26.000000
50%          31.000000
75%          38.000000
max          99.000000
Name: age, dtype: float64

AFTER FILTER, Zero Count in AGE column =  8335


age  duration_sec                start_time                  end_time  \
0   35         52185  2019-02-28 17:32:10.1450  2019-03-01 08:01:55.9750   
1    0         42521  2019-02-28 18:53:21.7890  2019-03-01 06:42:03.0560   

   start_station_id                                start_station_name  \
0              21.0  Montgomery St BART Station (Market St at 2nd St)   
1              23.0                     The Embarcadero at Steuart St   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.789625              -122.400811            13.0   
1               37.791464              -122.391034            81.0   

                 end_station_name  end_station_latitude  \
0  Commercial St at Montgomery St             37.794231   
1              Berry St at 4th St             37.775880   

   end_station_longitude  bike_id user_type  member_birth_year member_gender  \
0            -122.402923     4902  Customer               1984          Male   
1            -122.393170     2535  Customer                  0             0   

  bike_share_for_all_trip  
0                      No  
1                      No

Knowing the number of ZEROS in the ```AGE``` column may be important to consider when taking the mean or median (e.g., stats) of the dataframe. These ZEROS may need to be filtered/removed to ensure proper stats are completed with entries with acutal data.

Note we replaced all BLANKS with ZEROS <a href="#zero">earlier in the code.</a>

In [11]:
# help: https://cmdlinetips.com/2020/03/move-a-column-to-first-position-in-pandas-dataframe/
# moving age next to member_birth_year
col_name="age"
first_col = test.pop(col_name)
test.insert(14, col_name, first_col)
test.head(2)

duration_sec                start_time                  end_time  \
0         52185  2019-02-28 17:32:10.1450  2019-03-01 08:01:55.9750   
1         42521  2019-02-28 18:53:21.7890  2019-03-01 06:42:03.0560   

   start_station_id                                start_station_name  \
0              21.0  Montgomery St BART Station (Market St at 2nd St)   
1              23.0                     The Embarcadero at Steuart St   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.789625              -122.400811            13.0   
1               37.791464              -122.391034            81.0   

                 end_station_name  end_station_latitude  \
0  Commercial St at Montgomery St             37.794231   
1              Berry St at 4th St             37.775880   

   end_station_longitude  bike_id user_type  member_birth_year  age  \
0            -122.402923     4902  Customer               1984   35   
1            -122.393170     2535  Customer                  0    0   

  member_gender bike_share_for_all_trip  
0          Male                      No  
1             0                      No

In [12]:
# make test df into dfc again
dfc = test
dfc.head(2)

duration_sec                start_time                  end_time  \
0         52185  2019-02-28 17:32:10.1450  2019-03-01 08:01:55.9750   
1         42521  2019-02-28 18:53:21.7890  2019-03-01 06:42:03.0560   

   start_station_id                                start_station_name  \
0              21.0  Montgomery St BART Station (Market St at 2nd St)   
1              23.0                     The Embarcadero at Steuart St   

   start_station_latitude  start_station_longitude  end_station_id  \
0               37.789625              -122.400811            13.0   
1               37.791464              -122.391034            81.0   

                 end_station_name  end_station_latitude  \
0  Commercial St at Montgomery St             37.794231   
1              Berry St at 4th St             37.775880   

   end_station_longitude  bike_id user_type  member_birth_year  age  \
0            -122.402923     4902  Customer               1984   35   
1            -122.393170     2535  Customer                  0    0   

  member_gender bike_share_for_all_trip  
0          Male                      No  
1             0                      No

<a id='Q5'></a>

<font color='red'>

### ✔️ 5. <a href="#Q5">[Convert ```duration_sec``` to minutes, hours, or/and days]</a>

<a href="#todo">[Cleaning and Tidying To-do List]</a>

<a id='firstLast'></a>

In [13]:
# help: https://stackoverflow.com/questions/45579525/returning-a-dataframe-in-python-function/48368668
# preventing memory lost of df created in function

def firstLast(dataframe):
    cols = list(dataframe.columns)
    cols = [cols[-1]] + cols[:-1]
    dataframe = dataframe[cols]
    return dataframe

def timeConvertion(dataframe, seconds):
    dataframe['duration_min'] = dataframe[seconds] / 60
    dataframe = firstLast(dataframe) # save the dataframe within the function so it is not removed from memory
    dataframe['duration_hr'] = dataframe['duration_min'] / 60
    dataframe = firstLast(dataframe)
    return dataframe

test = dfc

test1 = timeConvertion(test, 'duration_sec')

test1.head(2)

<ipython-input-13-e9e6cb56d66f>:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe['duration_hr'] = dataframe['duration_min'] / 60


duration_hr  duration_min  duration_sec                start_time  \
0    14.495833    869.750000         52185  2019-02-28 17:32:10.1450   
1    11.811389    708.683333         42521  2019-02-28 18:53:21.7890   

                   end_time  start_station_id  \
0  2019-03-01 08:01:55.9750              21.0   
1  2019-03-01 06:42:03.0560              23.0   

                                 start_station_name  start_station_latitude  \
0  Montgomery St BART Station (Market St at 2nd St)               37.789625   
1                     The Embarcadero at Steuart St               37.791464   

   start_station_longitude  end_station_id                end_station_name  \
0              -122.400811            13.0  Commercial St at Montgomery St   
1              -122.391034            81.0              Berry St at 4th St   

   end_station_latitude  end_station_longitude  bike_id user_type  \
0             37.794231            -122.402923     4902  Customer   
1             37.775880            -122.393170     2535  Customer   

   member_birth_year  age member_gender bike_share_for_all_trip  
0               1984   35          Male                      No  
1                  0    0             0                      No

In [14]:
# assign prototype df to clean df
dfc = test1
dfc.head(2)

duration_hr  duration_min  duration_sec                start_time  \
0    14.495833    869.750000         52185  2019-02-28 17:32:10.1450   
1    11.811389    708.683333         42521  2019-02-28 18:53:21.7890   

                   end_time  start_station_id  \
0  2019-03-01 08:01:55.9750              21.0   
1  2019-03-01 06:42:03.0560              23.0   

                                 start_station_name  start_station_latitude  \
0  Montgomery St BART Station (Market St at 2nd St)               37.789625   
1                     The Embarcadero at Steuart St               37.791464   

   start_station_longitude  end_station_id                end_station_name  \
0              -122.400811            13.0  Commercial St at Montgomery St   
1              -122.391034            81.0              Berry St at 4th St   

   end_station_latitude  end_station_longitude  bike_id user_type  \
0             37.794231            -122.402923     4902  Customer   
1             37.775880            -122.393170     2535  Customer   

   member_birth_year  age member_gender bike_share_for_all_trip  
0               1984   35          Male                      No  
1                  0    0             0                      No

<a id='tidy-section'></a>

<font color='red'>

## TIDINESS ISSUES ADDRESSED -- note that ```df``` will change to ```dfc``` to indicate cleaned data

<a href="#clean">[Cleaning Data Requirements]</a> 
<a href="#todo">[Cleaning and Tidying To-do List]</a> 
<a href="#contents">[Table of Contents]</a>

<a id='T1'></a>

<font color='red'>

### ✔️ 1. <a href="#T1">[Create columns to break apart ```start_time``` to ```date``` and ```time```]</a>

<a href="#todo">[Cleaning and Tidying To-do List]</a>

The function ```firstLast``` can be found <a href="#firstLast">earlier in the code.</a>

In [15]:
def dateSeperator(dataframe, start, end):
    dataframe[end] = pd.to_datetime(dataframe[end])

    dataframe['end_hour'] = dataframe[end].dt.time
    dataframe = firstLast(dataframe)

    dataframe[start] = pd.to_datetime(dataframe[start])
    dataframe['start_hour'] = dataframe[start].dt.time
    dataframe = firstLast(dataframe)
    
    dataframe[end] = pd.to_datetime(dataframe[end])
    dataframe['end_day'] = dataframe[end].dt.date
    dataframe = firstLast(dataframe)
    
    dataframe[start] = pd.to_datetime(dataframe[start])
    dataframe['start_day'] = dataframe[start].dt.date
    dataframe = firstLast(dataframe)
    
    return dataframe

test = dfc

test1 = dateSeperator(test, 'start_time', 'end_time')

test1.head(2)

start_day     end_day       start_hour         end_hour  duration_hr  \
0  2019-02-28  2019-03-01  17:32:10.145000  08:01:55.975000    14.495833   
1  2019-02-28  2019-03-01  18:53:21.789000  06:42:03.056000    11.811389   

   duration_min  duration_sec              start_time                end_time  \
0    869.750000         52185 2019-02-28 17:32:10.145 2019-03-01 08:01:55.975   
1    708.683333         42521 2019-02-28 18:53:21.789 2019-03-01 06:42:03.056   

   start_station_id  ... end_station_id                end_station_name  \
0              21.0  ...           13.0  Commercial St at Montgomery St   
1              23.0  ...           81.0              Berry St at 4th St   

   end_station_latitude  end_station_longitude bike_id  user_type  \
0             37.794231            -122.402923    4902   Customer   
1             37.775880            -122.393170    2535   Customer   

   member_birth_year  age member_gender  bike_share_for_all_trip  
0               1984   35          Male                       No  
1                  0    0             0                       No  

[2 rows x 23 columns]

In [16]:
# recorganize data and check data type

# help: https://cmdlinetips.com/2020/03/move-a-column-to-first-position-in-pandas-dataframe/
col_name="start_time"
first_col = test1.pop(col_name)
test1.insert(4, col_name, first_col)

col_name="end_time"
first_col = test1.pop(col_name)
test1.insert(5, col_name, first_col)

test1.head(2)

start_day     end_day       start_hour         end_hour  \
0  2019-02-28  2019-03-01  17:32:10.145000  08:01:55.975000   
1  2019-02-28  2019-03-01  18:53:21.789000  06:42:03.056000   

               start_time                end_time  duration_hr  duration_min  \
0 2019-02-28 17:32:10.145 2019-03-01 08:01:55.975    14.495833    869.750000   
1 2019-02-28 18:53:21.789 2019-03-01 06:42:03.056    11.811389    708.683333   

   duration_sec  start_station_id  ... end_station_id  \
0         52185              21.0  ...           13.0   
1         42521              23.0  ...           81.0   

                 end_station_name  end_station_latitude  \
0  Commercial St at Montgomery St             37.794231   
1              Berry St at 4th St             37.775880   

   end_station_longitude bike_id  user_type  member_birth_year  age  \
0            -122.402923    4902   Customer               1984   35   
1            -122.393170    2535   Customer                  0    0   

  member_gender  bike_share_for_all_trip  
0          Male                       No  
1             0                       No  

[2 rows x 23 columns]

<a id='F1'></a>
<font color='blue'>

#### *FEATURE ENGINEERING: ✔️ 1. <a href="#todo">[Create column for ```day of week``` for the start/end days--to see what days bikes are rented the most]</a>*

*The function ```firstLast``` can be found <a href="#firstLast">earlier in the code.</a>*

In [17]:
# help: https://stackoverflow.com/questions/30222533/create-a-day-of-week-column-in-a-pandas-dataframe-using-python

def dayOfWeek(dataframe, start, end):
    
    dataframe[end] = pd.to_datetime(dataframe[end])
    dataframe['end_day_of_week'] = dataframe[end].dt.dayofweek
    dataframe = firstLast(dataframe)
    
    dataframe[start] = pd.to_datetime(dataframe[start])
    dataframe['start_day_of_week'] = dataframe[start].dt.dayofweek
    dataframe = firstLast(dataframe) 
    
    days = {0:'MON',1:'TUE',2:'WED',3:'THU',4:'FRI',5:'SAT',6:'SUN'}

    # look up day dictionary for day of week
    dataframe['end_day_of_week'] = dataframe['end_day_of_week'].apply(lambda x: days[x])
    dataframe['start_day_of_week'] = dataframe['start_day_of_week'].apply(lambda x: days[x])
    
    return dataframe


test2 = dayOfWeek(test1, 'start_time', 'end_time')

#print('distribution day of week\n', test2.groupby['start_day_of_week'].count())

count = test2.groupby('start_day_of_week')['start_time'].count().sort_values(ascending=False)
count1 = test2.groupby('end_day_of_week')['end_time'].count().sort_values(ascending=False)

print('START day of the week count\n', count)
print('\nEND day of the week count\n', count1)
test2.head(2)

START day of the week count
 start_day_of_week
THU    35176
TUE    31781
WED    29609
FRI    28962
MON    26807
SUN    15503
SAT    15377
Name: start_time, dtype: int64

END day of the week count
 end_day_of_week
THU    35164
TUE    31788
WED    29599
FRI    28962
MON    26814
SUN    15503
SAT    15385
Name: end_time, dtype: int64


start_day_of_week end_day_of_week   start_day     end_day       start_hour  \
0               THU             FRI  2019-02-28  2019-03-01  17:32:10.145000   
1               THU             FRI  2019-02-28  2019-03-01  18:53:21.789000   

          end_hour              start_time                end_time  \
0  08:01:55.975000 2019-02-28 17:32:10.145 2019-03-01 08:01:55.975   
1  06:42:03.056000 2019-02-28 18:53:21.789 2019-03-01 06:42:03.056   

   duration_hr  duration_min  ...  end_station_id  \
0    14.495833    869.750000  ...            13.0   
1    11.811389    708.683333  ...            81.0   

                 end_station_name end_station_latitude  end_station_longitude  \
0  Commercial St at Montgomery St            37.794231            -122.402923   
1              Berry St at 4th St            37.775880            -122.393170   

   bike_id  user_type member_birth_year  age  member_gender  \
0     4902   Customer              1984   35           Male   
1     2535   Customer                 0    0              0   

   bike_share_for_all_trip  
0                       No  
1                       No  

[2 rows x 25 columns]

In [18]:
print('\n----CLEANED STATE > Shape of File (row, column)---\n')
print(test2.shape)
print('\n----CLEANED STATE > Data Types---\n')
print(test2.dtypes)
print('\n----CLEANED STATE > Unique Values---\n')
print(test2.nunique())
print('\n----CLEANED STATE > Number of missing values---\n')
print(test2.isnull().sum())


----CLEANED STATE > Shape of File (row, column)---

(183215, 25)

----CLEANED STATE > Data Types---

start_day_of_week                  object
end_day_of_week                    object
start_day                          object
end_day                            object
start_hour                         object
end_hour                           object
start_time                 datetime64[ns]
end_time                   datetime64[ns]
duration_hr                       float64
duration_min                      float64
duration_sec                        int64
start_station_id                  float64
start_station_name                 object
start_station_latitude            float64
start_station_longitude           float64
end_station_id                    float64
end_station_name                   object
end_station_latitude              float64
end_station_longitude             float64
bike_id                             int64
user_type                          object
member_birth_yea

In [19]:
# reassign the df to dfc
dfc = test2
dfc.head(2)

start_day_of_week end_day_of_week   start_day     end_day       start_hour  \
0               THU             FRI  2019-02-28  2019-03-01  17:32:10.145000   
1               THU             FRI  2019-02-28  2019-03-01  18:53:21.789000   

          end_hour              start_time                end_time  \
0  08:01:55.975000 2019-02-28 17:32:10.145 2019-03-01 08:01:55.975   
1  06:42:03.056000 2019-02-28 18:53:21.789 2019-03-01 06:42:03.056   

   duration_hr  duration_min  ...  end_station_id  \
0    14.495833    869.750000  ...            13.0   
1    11.811389    708.683333  ...            81.0   

                 end_station_name end_station_latitude  end_station_longitude  \
0  Commercial St at Montgomery St            37.794231            -122.402923   
1              Berry St at 4th St            37.775880            -122.393170   

   bike_id  user_type member_birth_year  age  member_gender  \
0     4902   Customer              1984   35           Male   
1     2535   Customer                 0    0              0   

   bike_share_for_all_trip  
0                       No  
1                       No  

[2 rows x 25 columns]

<a id='T2'></a>

<font color='red'>

### ✔️ 2. <a href="#T2">[Create columns to break apart ```end_time``` to ```date``` and ```time```]</a>

<a href="#todo">[Cleaning and Tidying To-do List]</a>

*This was solved in <a href="#T2">Tidy Question 1 (above).</a>*

<hr size="5"/>
<h5><center>💾 Cleaning END ➜ Storing START 💾</center></h5>        
<hr size="5"/>

<a id='store'></a>
# 4. STORING

<font color=blue>
    

<a href="#store">[Table of Contents]</a>

In [20]:
# create csv from master dataframe - warning to user that opening these files "may" cause viewing software to crash
# help: https://www.guru99.com/python-check-if-file-exists.html

import os.path
from os import path
fileFlag1 = False # initilized as FALSE -- assuming .csv files have not be created from dataframes in this notebook


In [21]:
# help: https://www.geeksforgeeks.org/g-fact-41-multiple-return-values-in-python/
def createCSVmaster():
    # adding "1-" can for easier file handling in folder
    dfc.to_csv('master.csv', index=False)
    fileFlag1 = True
    print('master.csv File created ... Ready for user data analysis')
    return fileFlag1

if path.exists('master.csv') == True:
    print("File aready created! ... Ready for user data analysis")
elif fileFlag1 == False:
    fileFlag1 = createCSVmaster() # Assign returned tuple , execute file create assuming fileFlag1 is FALSE (has not be done)
else:
    assert path.exists('master.csv') == True, "You need to create the master.csv file"
    print("Files exists! ... Ready for user data analysis")

# could add button here for "confirmation" data was reviewed

master.csv File created ... Ready for user data analysis


<hr size="5"/>
<h5><center>🗺️ Storing END ➜ Data Exploration START 🗺️</center></h5>        
<hr size="5"/>

<a id='explore'></a>
# 5. DATA EXPLORATION

<font color=blue>
    
<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br>
--- <a href="#preliminary1">[Preliminary Review]</a>
--- <a href="#univariate1">[Univariate Exploration]</a>
--- <a href="#bivariate1">[Bivariate Exploration]</a>
--- <a href="#multivariate1">[Multivariate Exploration]</a> 

<a href="#discussion">[#6 Discussion]</a><br>
--- <a href="#preliminary2">[Preliminary Review]</a>
--- <a href="#univariate2">[Univariate Exploration]</a>
--- <a href="#bivariate2">[Bivariate Exploration]</a>
--- <a href="#multivariate2">[Multivariate Exploration]</a> 

<a id='preliminary1'></a>
## 5.1 Exploration - Preliminary Review

<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br> 

<a href="#discussion">[#6 Discussion]</a><br>
--- <a href="#preliminary2">[6.1 Discussion - Preliminary Review]</a>

Review of the distribution of the following variables
1. age
2. Time and Days Bikes Used

In [22]:
# help: https://mode.com/example-gallery/python_histogram/

<a id='univariate1'></a>
## 5.2 Exploration - Univariate Exploration
<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br> 

<a href="#discussion">[#6 Discussion]</a><br>
--- <a href="#univariate2">[6.2 Discussion - Univariate Exploration]</a>

<a id='bivariate1'></a>
## 5.3 Exploration - Bivariate Exploration
<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br>

<a href="#discussion">[#6 Discussion]</a><br>
--- <a href="#bivariate2">[6.3 Discussion - Bivariate Exploration]</a>

<a id='multivariate1'></a>
## 5.4 Exploration - Multivariate Exploration
<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br> 

<a href="#discussion">[#6 Discussion]</a><br>
--- <a href="#multivariate2">[6.4 Discussion - Multivariate Exploration]</a> 

<hr size="5"/>
<h5><center>📊 Data Exploration END ➜ Discussion START 📊</center></h5>        
<hr size="5"/>

<a id='discussion'></a>
# 6. DISCUSSION

<font color=blue>
    
<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br>
--- <a href="#preliminary1">[Preliminary Review]</a>
--- <a href="#univariate1">[Univariate Exploration]</a>
--- <a href="#bivariate1">[Bivariate Exploration]</a>
--- <a href="#multivariate1">[Multivariate Exploration]</a> 

<a href="#discussion">[#6 Discussion]</a><br>
--- <a href="#preliminary2">[Preliminary Review]</a>
--- <a href="#univariate2">[Univariate Exploration]</a>
--- <a href="#bivariate2">[Bivariate Exploration]</a>
--- <a href="#multivariate2">[Multivariate Exploration]</a> 

<a id='preliminary2'></a>
## 6.1 Discussion - Preliminary Wrangling
<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br>
--- <a href="#preliminary1">[5.1 Exploration - Preliminary Review]</a>

<a href="#discussion">[#6 Discussion]</a><br>
    
### What is the structure of your dataset?

There are 53,940 diamonds in the dataset with 10 features (carat, cut, color, clarity, depth, table, price, x, y, and z). Most variables are numeric in nature, but the variables cut, color, and clarity are ordered factor variables with the following levels.

(worst) ——> (best) <br>
cut: Fair, Good, Very Good, Premium, Ideal <br>
color: J, I, H, G, F, E, D <br>
clarity: I1, SI2, SI1, VS2, VS1, VVS2, VVS1, IF

### What is/are the main feature(s) of interest in your dataset?

I'm most interested in figuring out what features are best for predicting the price of the diamonds in the dataset.

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

I expect that carat will have the strongest effect on each diamond's price: the larger the diamond, the higher the price. I also think that the other big "C"s of diamonds: cut, color, and clarity, will have effects on the price, though to a much smaller degree than the main effect of carat.

<a id='univariate2'></a>
##  6.2 Discussion - Univariate Exploration

<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br>
--- <a href="#univariate1">[5.2 Exploration - Univariate Exploration]</a>

<a href="#discussion">[#6 Discussion]</a><br>

I'll start by looking at the distribution of the main variable of interest: price.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

The price variable took on a large range of values, so I looked at the data using a log transform. Under the transformation, the data looked bimodal, with one peak between \$500 and \$1000, and another just below \$5000.

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

When investigating the x, y, and z size variables, a number of outlier points were identified. Overall, these points can be characterized by an inconsistency between the recorded value of depth, and the value that would be derived from using x, y, and z. For safety, all of these points were removed from the dataset to move forwards.

<a id='bivariate2'></a>
## 6.3 Discussion - Bivariate Exploration

<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br>
--- <a href="#bivariate1">[5.3 Exploration - Bivariate Exploration]</a>

<a href="#discussion">[#6 Discussion]</a><br>

To start off with, I want to look at the pairwise correlations present between features in the data.

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

Price had a surprisingly high amount of correlation with the diamond size, even before transforming the features. An approximately linear relationship was observed when price was plotted on a log scale and carat was plotted with a cube-root transform. The scatterplot that came out of this also suggested that there was an upper bound on the diamond prices available in the dataset, since the range of prices for the largest diamonds was much narrower than would have been expected, based on the price ranges of smaller diamonds.

There was also an interesting relationship observed between price and the categorical features. For all of cut, color, and clarity, lower prices were associated with increasing quality. One of the potentially major interacting factors is the fact that improved quality levels were also associated with smaller diamonds. This will have to be explored further in the next section.

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

Expected relationships were found in the association between the 'x', 'y', and 'z' measurements of diamonds to the other linear dimensions as well as to the 'carat' variable. A small negative correlation was observed between table size and depth, but neither of these variables show a strong correlation with price, so they won't be explored further. There was also a small interaction in the categorical quality features. Diamonds of lower clarity appear to have slightly better cut and color grades.

<a id='multivariate2'></a>
## 6.4 Discussion - Multivariate Exploration

<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br>
--- <a href="#multivariate1">[5.4 Exploration - Multivariate Exploration]</a> 

<a href="#discussion">[#6 Discussion]</a><br>

The main thing I want to explore in this part of the analysis is how the three categorical measures of quality play into the relationship between price and carat.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

I extended my investigation of price against diamond size in this section by looking at the impact of the three categorical quality features. The multivariate exploration here showed that there indeed is a positive effect of increased quality grade on diamond price, but in the dataset, this is initially hidden by the fact that higher grades were more prevalent in smaller diamonds, which fetch lower prices overall. Controlling for the carat weight of a diamond shows the effect of the other C's of diamonds more clearly. This effect was clearest for the color and clarity variables, with less systematic trends for cut.

### Were there any interesting or surprising interactions between features?

Looking back on the point plots, it doesn't seem like there's a systematic interaction effect between the three categorical features. However, the features also aren't fully independent. But it is interesting in something like the 1-carat plot for prices against cut and clarity, the shape of the 'cut' dots is fairly similar for the SI2 through VVS2 clarity levels.

<hr size="5"/>
<h5><center>🍻 Discussion END ➜ Conclusion START 🍻</center></h5>        
<hr size="5"/>

<a id='conclusion'></a>
# 7. CONCLUSION


<font color=blue>
    
<a href="#contents">[Table of Contents]</a>

<a href="#explore">[#5 Data Exporation]</a><br>
--- <a href="#preliminary1">[Preliminary Review]</a>
--- <a href="#univariate1">[Univariate Exploration]</a>
--- <a href="#bivariate1">[Bivariate Exploration]</a>
--- <a href="#multivariate1">[Multivariate Exploration]</a> 

<a href="#discussion">[#6 Discussion]</a><br>
--- <a href="#preliminary2">[Preliminary Review]</a>
--- <a href="#univariate2">[Univariate Exploration]</a>
--- <a href="#bivariate2">[Bivariate Exploration]</a>
--- <a href="#multivariate2">[Multivariate Exploration]</a> 

<hr size="5"/>
<h5><center>🏁 Conclusion END -- Project FINALE 🏁</center></h5>        
<hr size="5"/>

# End of Data Project!

Made with ❤️ by Jhon!

Further impovements include...
1. Validation statements using ```assert``` clauses to confirm that data manipuation is correct during cleaning stage
2. More plots exploring data

<a href="#contents">[Table of Contents]</a>